### ISSUE TITLE
[Deployment] Add a cookbook about deploying Symfony with Docker

### ISSUE THREAD

**Issue by javiereguiluz** [symfony/symfony-docs#4055]
One of the most requested topics by Symfony developers is how to deploy their applications. The cookbook already contains a new **Deployment** section with articles about how to deploy Symfony on Windows Azure and Heroku.

In addition to those articles, we need a new article about **how to deploy Symfony with Docker**. Instead of duplicating the blog posts and other resources published about this, the idea is to publish a short article which becomes the the _canonical resource_ about Symfony + Docker.

Nobody is yet assigned to this task. If you want to give it a try, please tell us in the comments below. That way we'll coordinate better and we won't duplicate efforts.

**Comment by theofidry** [comment]
+1 would be great to have it on the official doc.

**Comment by mlpo** [comment]
+1, I agree, it would be really great

**Comment by WouterJ** [comment]
If someone has succesfully set up a Symfony app on docker, I would like to have an overview of the things you had to do to make it working. This way, we can bootstrap such an article. (you don't have to write the article, just an overview of the required steps and additional things specific to docker are great!)

**Comment by javiereguiluz** [comment]
@WouterJ this is the best presentation that I'm aware of about this and it explains everything with great detail: https://speakerdeck.com/ubermuda/a-multi-container-symfony2-setup-with-docker

**Comment by ivoba** [comment]
Would this aim at a production container or also at a development container?

**Comment by WouterJ** [comment]
Another set of docker files: https://github.com/samsch/docker-symfony

I still have no experience with Docker, so it would be great if someone else can start the bootstrapping of the article.

**Comment by pschultz** [comment]
Writing about the actual *deployment* of Docker containers, as in executing the containers, is out of scope of a Cookbook article because there are so many options. Bare metal, Swarm, CoreOS Fleet, ECS, Google Container Engine etc.  etc. Instead, an article should explain how to build images for Symfony applications that can be deployed in most of those environments.

@WouterJ asked for an experience report by someone that has setup Symfony with Docker. At classmarkets (that's my employer), we're running dockered PHP applications in production for a while now -- Symfony and otherwise. I thought I'd share how we do that.

> N.B. this is about production images only and doesn't consider development at all.

If we apply what I am about to explain to symfony-demo, we end up with something like [this diff][diff].

As with "Deploying to Heroku Cloud" the logger should stream to php://stderr. And due to the ephemeral nature of Docker containers most considerations for distributed deployments apply, even if running just a single app instance. For example, don't put sessions on disk or they will be gone after restart.

We also want to be able to set parameters through environment variables. We do that with a php file that is included after parameters.yml, see the diff.


I've submitted our fork to the Docker Hub, so you can actually run symfony-demo on [http://localhost:8080/](http://localhost:8080/) with this docker-compose.yml:

```yml
app:
    image: classmarkets/symfony-demo:docker

nginx:
    image: nginx:mainline-alpine
    command: nginx -g 'daemon off; error_log stderr notice;'
    links:
        - app
    volumes_from:
        - app:ro
    ports:
        - "8080:80"
```

I don't mean to imply that docker-compose is the way to go about deployments.  I'm just using it to show how the application is intended to be run.


The tricky part is of course to write the actual Dockerfile(s), and there is certainly more than one way to do it. What follows is the rational for our way to skin a cat. (We're not really skinning cats.)


For each app instance, we run two containers alongside each other. One for php-fpm and one for nginx.


I frequently see sample setups that put the application code into a third, volume-only container. We don't. The whole point of Docker is to build self-contained, executable artifacts. We do not separate our application code from its runtime environment. How do you make sure that a php-only image contains all the extensions that the code-only image requires?  And that the ini settings match up? The answer is, you don't. Unless you always build and tag them together, at which point you might as well build just one image, create one less container, and safe yourself the trouble of inadvertently breaking docker links.


To make the PHP container useful we run a cooperating nginx container that translates between HTTP and FastCGI. Nginx needs to know a few things about the PHP setup. Which port is php-fpm listening on? Where is the webroot? How is the index file named? Do I have to rewrite requests? If so, how? We have to configure that somewhere.

We don't want to build a specialized nginx image for each app because that suffers from the exact same problem I mentioned above about code-only images.  The webserver configuration changes over time and has to stay compatible with the application code. At some point we added version query strings to all asset urls in a particular application (e.g. "/foo.png?v1.3") so that nginx could set aggressive Cache-Control headers for all static assets. This is an example for two changes that should really be deployed together and, maybe more importantly, rolled back together. Independent nginx and php images don't help with that.

That's why we keep the nginx configuration in the application image. Not all of it, just the parts specific to the app: the virtual host configuration.  Then, at runtime, we share that configuration snippet with the nginx container using the --volumes-from option.

Notice that the nginx image in the docker-compose.yml above is a stock image, straight from the hub. There's nothing special about it. The only requirement is that it has to include /etc/nginx/conf.d/* or equivalent, and have fcgi support, obviously. There is no Dockerfile for this nginx image in the repository. This is how Docker was designed to work. One Dockerfile per app, at the repository root.


The downside is that there's one nginx for each php-fpm process which isn't exactly economical; one nginx could easily proxy dozens if not hundreds of php-fpms. On the other hand, nginx is so lightweight that it doesn't matter to us in practice.

Summing up: The article should explain how to author a Dockerfile. That Dockerfile should be buildable in a wide range of environments and produce an image that can be run in a wide range of environments. [The diff][diff] shows one such Dockerfile.

[diff]: https://github.com/symfony/symfony-demo/compare/master...classmarkets:docker

**Comment by pschultz** [comment]
Maybe I should have mentioned Apache + mod_php. I was just assuming no one does that anymore. Using Apache would make the setup far less complex, since it requires only one running container. Our nginx image that we have anyway offers more flexibilty so we just run php-fpm.

That being said, in context of a cookbook article Apache+mod_php would be much easier to explain. The question is, if it adds enough value. That really depends on what the article is supposed to be achieve. Is it about building images that are useful in practice, in professional environments? In my experience mod_php doesn't qualify anymore but I may be wrong about that.

**Comment by sagikazarmark** [comment]
If it matters, I've setup a development environment using Symfony and Docker:

https://github.com/webplates/symfony-standard

See the wiki for details.

**Comment by webberig** [comment]
There are actually plenty of resources out there on how to run your Symfony app inside a docker container for development. In this case you map your local folder to a docker volume.

But this is not how you would publish your app as a production-ready docker image, and I can't seem to find much resources on how you woould do that.

I'm having a bit of success for a project, but there are many issues to work out. For starters my docker image is huge and I can't find out why. Secondly, I' stuck with logging:
http://stackoverflow.com/questions/38499825/symfony-logs-to-stdout-inside-docker-container

It would be great if a docker expert hooked up with a Symfony dev and made a baby together, metaforically speaking! :-D

**Comment by ro0NL** [comment]
For the record, my home setup lives at https://github.com/ro0NL/workdir (a generic dev environment for multiple projects).

- ssh access (this is extremely convenient btw, my host user is also the guest user; so no permission hassle whatsoever)
- composer / symfony installer / xdebug available out-of-the-box
- dockerized commands (i.e. run composer from the host)

**Comment by cordoval** [comment]
@timglabisch you need to quote your envs "${PASSWORD}" otherwise hell unleashes on kubernetes and i wonder where are you warming up cache on the build or on prod

**Comment by javiereguiluz** [comment]
I'm closing this issue because we haven't fixed it in more than 5 years. Either there is not enough interest in the community or this is too hard to do or not really needed anymore.

That's fine. We don't have to make all our ideas a reality. Let's focus on more important things. Thanks!
